cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -298,8 +291,6 @@ public KafkaAssemblyOperator(Vertx vertx, PlatformFeaturesAvailability pfa,
                 .compose(state -> state.kafkaManualPodCleaning())
                 .compose(state -> state.kafkaNetPolicy())
                 .compose(state -> state.kafkaManualRollingUpdate())
-                .compose(state -> state.kafkaVersionChangeCheck())
-                .compose(state -> state.kafkaVersionChange(false))
```

Comment:
The original idea with having the version change roll early in the kafka part of the reconciliation is the operator hasn't done anything to the kafka pods at this point, so it doesn't needed to make any assumptions about how the image behaves/what it expects, which is relevant when the operator is upgraded and does the forced operand upgrade.

By deferring the forced operand upgrade until after we've done things like changed the networking and updated the ancillary CM there is a higher risk that one of the brokers suffers a spontaneous failure, restarts and doesn't see compatible config (because the image in the sts hasn't been changed yet). The KafkaConfig algorithm should be enough because it always tries to roll brokers which aren't ready first. So any broker which had spontaneously restarted would restart first and because it's then using the right image it should be able to start. So I _think_ this is a recoverable error, assuming that all such failures result in an unready broker.

Reply:
Yeah, I know that was the original motivation. But it was also the part which caused all the issues because it required knowledge of prior versions. In theory yes, something can happen and break the cluster. But I think the time is fairly short and the benefits from using the regular rolling update overweight the negatives of having to fight with the prior knowledge.